zebra-network/src/peer_set/candidate_set/tests/vectors.rs

```diff
@@ -0,0 +1,108 @@
+use std::net::{IpAddr, SocketAddr};
+
+use chrono::{DateTime, Duration, Utc};
+
+use super::super::validate_addrs;
+use crate::types::{MetaAddr, PeerServices};
+
+/// Test that offset is applied when all addresses have `last_seen` times in the future.
+#[test]
+fn offsets_last_seen_times_in_the_future() {
+    let last_seen_limit = Utc::now();
+
+    let input_peers = mock_gossiped_peers(vec![
+        last_seen_limit + Duration::minutes(30),
+        last_seen_limit + Duration::minutes(15),
+        last_seen_limit + Duration::minutes(45),
+    ]);
+
+    let validated_peers: Vec<_> = validate_addrs(input_peers, last_seen_limit).collect();
+
+    let expected_offset = Duration::minutes(45);
+    let expected_peers = mock_gossiped_peers(vec![
+        last_seen_limit + Duration::minutes(30) - expected_offset,
+        last_seen_limit + Duration::minutes(15) - expected_offset,
+        last_seen_limit + Duration::minutes(45) - expected_offset,
+    ]);
+
+    assert_eq!(validated_peers, expected_peers);
+}
+
+/// Test that offset is not applied if all addresses have `last_seen` times that are in the past.
+#[test]
+fn doesnt_offset_last_seen_times_in_the_past() {
+    let last_seen_limit = Utc::now();
+
+    let input_peers = mock_gossiped_peers(vec![
+        last_seen_limit - Duration::minutes(30),
+        last_seen_limit - Duration::minutes(45),
+        last_seen_limit - Duration::days(1),
+    ]);
+
+    let validated_peers: Vec<_> = validate_addrs(input_peers.clone(), last_seen_limit).collect();
+
+    let expected_peers = input_peers;
+
+    assert_eq!(validated_peers, expected_peers);
+}
+
+/// Test that offset is applied to all the addresses if at least one has a `last_seen` time in the
+/// future.
+///
+/// Times that are in the past should be changed as well.
+#[test]
+fn offsets_all_last_seen_times_if_one_is_in_the_future() {
+    let last_seen_limit = Utc::now();
+
+    let input_peers = mock_gossiped_peers(vec![
+        last_seen_limit + Duration::minutes(55),
+        last_seen_limit - Duration::days(3),
+        last_seen_limit - Duration::hours(2),
+    ]);
+
+    let validated_peers: Vec<_> = validate_addrs(input_peers, last_seen_limit).collect();
+
+    let expected_offset = Duration::minutes(55);
+    let expected_peers = mock_gossiped_peers(vec![
+        last_seen_limit + Duration::minutes(55) - expected_offset,
+        last_seen_limit - Duration::days(3) - expected_offset,
+        last_seen_limit - Duration::hours(2) - expected_offset,
+    ]);
+
+    assert_eq!(validated_peers, expected_peers);
+}
+
+/// Test that offset is not applied if the most recent `last_seen` time is equal to the limit.
+#[test]
+fn doesnt_offsets_if_most_recent_last_seen_times_is_exactly_the_limit() {
+    let last_seen_limit = Utc::now();
```

Comment:
👍 

Thanks for using a single time here.

I've seen tests fail because we call `now()` twice, and the second ticks over during the test.